compiler/rustc_session/src/search_paths.rs

```diff
@@ -10,9 +13,40 @@ use crate::filesearch::make_target_lib_path;
 pub struct SearchPath {
     pub kind: PathKind,
     pub dir: PathBuf,
-    pub files: Vec<SearchPathFile>,
+    pub files: FilesIndex,
 }
 
+/// [FilesIndex] contains paths that can be efficiently looked up with (prefix, suffix) pairs.
+#[derive(Clone, Debug)]
+pub struct FilesIndex(BTreeMap<Arc<str>, Arc<SearchPathFile>>);
```

Comment:
Since this is built once and then reused, it might make sense to use `Vec<SearchPathFile>` or similar and sorting it + [partition_point](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.partition_point) for lookup. I'd expect that to be both faster on lookup and might allow skipping the Arcs?